app/v2/app.go

```diff
@@ -0,0 +1,248 @@
+func New(c Config) *App {
+	log.Trace().Msg("[app] Creating new app")
+
+	readinessProbeGroup := NewProbeGroup("readiness")
+	healthnessProbeGroup := NewProbeGroup("healthness")
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	app := &App{
+		Ready:               readinessProbeGroup,
+		Healthy:             healthnessProbeGroup,
+		mainLoopCtx:         ctx,
+		cancelMainLoopCtx:   cancel,
+		gracePeriod:         c.App.Shutdown.GracePeriod,
+		shutdownTimeout:     c.App.Shutdown.Timeout,
+		mainReadinessProbe:  readinessProbeGroup.MustNewProbe("fkit/app", false),
+		mainHealthnessProbe: healthnessProbeGroup.MustNewProbe("fkit/app", true),
+	}
+
+	app.startAdminServer(c)
+
+	if c.App.Trace.Exporter != "" {
+		trace.SetupTrace(c.App.Trace)
+	}
+
+	return app
+}
+
+func (app *App) startAdminServer(c Config) {
+	if !c.App.AdminServer.Enabled {
+		return
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	mux.Handle("/healthy", &app.Healthy)
+	mux.Handle("/ready", &app.Ready)
+
+	if c.App.AdminServer.With.DebugURLs {
+		mux.HandleFunc("/debug/pprof/", http.HandlerFunc(pprof.Index))
+		mux.HandleFunc("/debug/pprof/cmdline", http.HandlerFunc(pprof.Cmdline))
+		mux.HandleFunc("/debug/pprof/profile", http.HandlerFunc(pprof.Profile))
+		mux.HandleFunc("/debug/pprof/symbol", http.HandlerFunc(pprof.Symbol))
+		mux.HandleFunc("/debug/pprof/trace", http.HandlerFunc(pprof.Trace))
+		mux.HandleFunc("/debug/dump/goroutines", dumpGoroutines)
+		mux.HandleFunc("/debug/dump/memory", dumpMemProfile)
+		mux.HandleFunc("/debug/dump/memstats", dumpMemStats)
+	}
+
+	server := http.Server{
+		Addr:              ":" + c.App.AdminServer.Port,
+		Handler:           mux,
+		ReadHeaderTimeout: 60 * time.Second,
+	}
+	go func() {
+		err := server.ListenAndServe()
+		if err != nil {
+			log.Fatal().Err(err).Msg("Failed to start admin server.")
+		}
+	}()
+}
+
+// Shutdown calls all shutdown methods, in order they were added.
+func (app *App) Shutdown(ctx context.Context) (err error) {
+	log.Trace().Int("shutdown_handlers", app.shutdownHandlers.Len()).Msg("[app] Starting graceful shutdown.")
+
+	app.cancelMainLoopCtx()
+
+	const op = errors.Op("app.App.Shutdown")
+
+	if app.shutdownTimeout > 0 {
+		log.Trace().Dur("shutdown_timeout", app.shutdownTimeout).Msg("[app] Configuring a timeout for the shutdown.")
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, app.shutdownTimeout)
+		defer cancel()
+	}
+
+	select {
+	case <-ctx.Done():
+		err = errors.E(op, "shutdown deadline has been reached")
+	case err = <-app.shutdownAllHandlers(ctx):
+	}
+	if err != nil {
+		log.Trace().Err(err).Msg("[app] Graceful shutdown failed.")
+		return errors.E(op, err)
+	}
+
+	log.Trace().Msg("[app] Graceful shutdown finished successfully.")
+	return nil
+}
+
+func (app *App) shutdownAllHandlers(ctx context.Context) chan error {
+	const op = errors.Op("shutdownAllHandlers")
+	done := make(chan error, 1)
```

Comment:
You are right. Only the first write matters because there is only one goroutine here. The channel is used to pass the error asynchronously.
But I've just realized we have a bug here. It's missing a return statement.